Enable TensorExpressions to internally split up equations

macedo22 opened the pull request:
## Proposed changes

This PR greatly improves the runtime of `TensorExpression`s, especially those that have many terms and/or tensor operations.

Benchmarking found two sources to be cause of a lot of slowdown:
- breadth in the tree
- large `DataVector` expressions (i.e. massive one-liners generated by equations with many operations)

The changes done in this PR address these two points by:
- reorganizing the expression tree to have more depth and less breadth
- "splitting up" TEs internally by evaluating subtrees when there are a large number of `DataVector` operations

Because the changes in this PR are sweeping (all TEs affected) and there is more cognitive complexity than before, I leaned more on over-documenting. When reviewing, **I recommend first reading the new documentation I've added to `TensorExpression.hpp`**, which defines some new terms I'm defining, as well as how, at a high level, the tree recursion works with the splitting.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Contract all index pairs in one TensorContract instead of nesting

This changes the contraction expression tree from being short and wide to long and skinny. This was done because it was found that breadth in TensorExpression trees typically corresponds to worse runtime.

macedo22 committed:
Count # tensor ops in TE subtrees, reorder commutative binary ops

This changes a TensorExpression tree from being short and wide to longer and skinnier. This was done because it was found that breadth in TensorExpression trees typically corresponds to worse runtime.

macedo22 committed:
Enable TensorExpressions to internally split up equations

When a TensorExpression is large enough, it can be faster to compute subexpressions of the full equation than to evaluate the equation as one big one-liner.

macedo22 committed:
Add minor optimization for TE multi-index transformations

nilsvu commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -40,289 +40,715 @@ using indices_contractible = std::bool_constant<
+  std::array<bool, NumUncontractedIndices> index_mapping_set{};
+  for (size_t i = 0; i < NumUncontractedIndices; i++) {
+    gsl::at(index_mapping_set, i) = false;
+  }
```
(optional) I think these three lines can be safely deleted. The std::array is default-initialized to false.

macedo22 replied:
I ended up switching the for loop to a call to `make_array<NumUncontractedIndices, bool>(false)`. I tried finding an answer as to whether or not default-initialization of bools was `0` or unknown and I found [this page](https://en.cppreference.com/w/cpp/language/default_initialization). I'm not 100% clear on the different storage types and much less so about threads, but I think this note that it has at the bottom means that we can't count on `bool` being initialized to `0`:

> Default initialization of non-class variables with automatic and dynamic storage duration produces objects with indeterminate values (static and thread-local objects get [zero initialized](https://en.cppreference.com/w/cpp/language/zero_initialization))

nilsvu replied:
`make_array` is good 👍

nilsvu commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -40,289 +40,715 @@ using indices_contractible = std::bool_constant<
+    gsl::at(index_mapping_set, i) = false;
+  }
+
+  // Index of `contracted_index_pair_positions` that we're current'y assigning
```
typo: currently

nilsvu commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -40,289 +40,715 @@ using indices_contractible = std::bool_constant<
+  // Index of `index_transformation` that we're currently assigning
+  size_t not_contracted_map_index_to_assign =
+      NumUncontractedIndices - NumContractedIndexPairs * 2 - 1;
+  for (size_t i = NumUncontractedIndices - 1; i < NumUncontractedIndices; i--) {
```
I don't understand the `i < NumUncontractedIndices` clause. Shouldn't it be `i > 0`? If yes, why does this work at the moment? If no, could you explain a bit in a code comment? Please also add a comment why you are iterating backwards here.

macedo22 replied:
The reason I iterated backwards here is for a pretty weak hunch for a specific case, but it took quite a few words to describe :sweat_smile:, I apologize. I didn't benchmark backwards vs forwards for this, and I don't think there is any reason why I couldn't iterate forward, instead, so let me know if that's preferred.

nilsvu replied:
Thanks for adding the comments 👍

nilsvu commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -40,289 +40,715 @@ using indices_contractible = std::bool_constant<
+
+/// \brief Computes type information for the tensor expression that results from
+/// a contraction, as well as information internally useful for carrying out the
+/// contractin
```
contraction

nilsvu commented on src/DataStructures/Tensor/Expressions/TensorAsExpression.hpp:
```diff
@@ -179,13 +179,32 @@ struct TensorAsExpression<Tensor<X, Symm, IndexList<Indices...>>,
+  static constexpr size_t num_ops_right_child = 0;
+  /// The total number of arithmetic tensor operations done in this expression's
+  /// whole subtree, which is 0 because this is a leaf expression
+  static constexpr size_t num_ops_subtree = 0;
```
Would it be useful to add tests that check the `num_ops_subtree` of some example expressions is what you would expect?

macedo22 replied:
Thanks for the suggestion! I've added tests like you've described to `Test_AddSubtract.cpp`, `Test_Contract.cpp`, `Test_Divide.cpp`, `Test_Negate.cpp`, `Test_Product.cpp`, and `Test_SquareRoot.cpp` in functions called `test_tensor_ops_properties()`. `TensorAsExpression` and `NumberAsExpression` are indirectly also tested through their usage in those tests.

macedo22 replied:
Now that I think about it, I think the `get_used_for_size` functions I've added to each derived TE class probably isn't needed anymore for the same reason - not unless there is an expected performance difference between the following two cases:

(a)
```
tnsr::i<DataVector> R(used_for_size);
get<0>(R) = ...;
get<1>(R) = ...;
get<2>(R) = ...;
```

(b)
```
tnsr::i<DataVector> R{};
get<0>(R) = ...;
get<1>(R) = ...;
get<2>(R) = ...;
```

Do either you or @nilsdeppe know if (a) or (b) is generally preferred? If not, it's something I could look into

nilsvu commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -353,76 +368,94 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
       num_ops_left_child + num_ops_right_child + 1;
 
+  // === Properties for splitting up subexpressions along the primary path ===
+  // These defintiions only have meaning if this expression actually ends up
```
typo: definitions

macedo22 replied:
Fixed this in the other files that have this typo, as well

nilsvu commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -22,12 +24,231 @@ struct Expression {};
+/// wherever reasonable
+/// - Minimize breadth in the tree where possible, as benchmarking inner
+/// products has shown that increased tree breadth can cause slower runtimes.
+/// In addition, more breadth means a descreased ability to split up the tree
```
decreased

nilsvu commented on tests/Unit/DataStructures/Tensor/Expressions/Test_MixedOperations.cpp:
```diff
@@ -129,7 +129,7 @@ void test_case1(const DataType& used_for_size,
       TensorExpressions::evaluate<ti_a>(R(ti_a, ti_b) * S(ti_B) + G(ti_a) -
                                         H(ti_b, ti_a, ti_B) * T());
-  result_tensor_type actual_result_tensor_filled{};
+  result_tensor_type actual_result_tensor_filled(used_for_size);
```
Why is this necessary now? I may have missed this, but is the general policy that the result buffer must be sized correctly on entry, or will `TE::evaluate` resize it? Is this documented?

macedo22 replied:
I think this was a vestige of a time during the design process when this necessary, but it no longer is, so I've removed it. TEs should work whether the input LHS tensor has been sized or not because it uses assignment like `lhs_component = get(result of some subtree)`, which should assign the LHS component using a sized RHS `Tensor` component

macedo22 replied:
Now that I think about it, I think the `get_used_for_size` functions I've added to each derived TE class probably aren't needed anymore for the same reason - not unless there is an expected performance difference between the following two cases:

(a)
```
tnsr::i<DataVector> R(used_for_size);
get<0>(R) = ...;
get<1>(R) = ...;
get<2>(R) = ...;
```

(b)
```
tnsr::i<DataVector> R{};
get<0>(R) = ...;
get<1>(R) = ...;
get<2>(R) = ...;
```

Do either you or @nilsdeppe know if (a) or (b) is generally preferred? If not, it's something I could look into

nilsvu replied:
I think (b) might actually be slightly better because (a) calls `make_array<size, DataVector>(used_for_size)` which copies the vector `size` times. Therefore, if you can make the implementation simpler with (b) then I'd say do it.

macedo22 replied:
Addressed by new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/b1566dabe4805a70c10e5a17c698a07774d3fbdb

nilsvu commented on src/DataStructures/Tensor/Expressions/TensorIndexTransformation.hpp:
```diff
@@ -151,11 +151,10 @@ transform_multi_index(
   for (size_t i = 0; i < NumIndicesOut; i++) {
     gsl::at(output_multi_index, i) =
-        (gsl::at(tensorindex_transformation, i) ==
+        (gsl::at(tensorindex_transformation, i) !=
```
how does this optimization work? Add a code comment?

macedo22 replied:
I added a comment on this. Note that I didn't benchmark before vs after this. I just made this swap because it's my understanding that with control flow, in general, we want to put the more likely cases as the `if` instead of the `else`. I don't know if it has a meaningful impact, as this is one of the many small changes I made in the overall redesign. If you prefer I not make this change without benchmarking, let me know and I will remove this change for now!

nilsvu replied:
I have no idea if this is faster, but I don't mind the change 🤷‍♂️ Thanks for adding the comment.

nilsvu reviewed (changes requested):
I can't say that I understood everything, but the fact that the tests didn't change gives me confidence that at least things are not broken :)

nilsvu commented:
Please squash changes directly.

nilsvu commented:
You got a CI error for [clang-11 Debug](https://github.com/sxs-collaboration/spectre/runs/5386348239?check_suite_focus=true):

```
make[3]: *** [src/Evolution/Systems/Ccz4/CMakeFiles/Ccz4.dir/build.make:147: src/Evolution/Systems/Ccz4/CMakeFiles/Ccz4.dir/Ricci.cpp.o] Error 255
make[2]: *** [CMakeFiles/Makefile2:17856: src/Evolution/Systems/Ccz4/CMakeFiles/Ccz4.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
```

and also for [clang-8 Debug](https://github.com/sxs-collaboration/spectre/runs/5386348032?check_suite_focus=true):

```
clang: error: unable to execute command: Killed
```

I think the latter is clang running out of memory, the former maybe as well. This worries me a bit. I also noticed that building unit tests on CI took quite a bit longer than, e.g., https://github.com/sxs-collaboration/spectre/runs/5382670412?check_suite_focus=true. Do you have any indication that the changes you made here increased compile time significantly? The compile time difference could be due to caching, but it would be good to make sure that compile time and compiler memory usage are reasonable for these changes.

macedo22 commented:
@nilsvu I've only been building the TE tests locally and I hadn't noticed any slowdown with the compile time, but I also haven't been looking at specific times or RAM usage. It looks like maybe building `Evolution/Systems/Ccz4/Ricci.cpp` is problematic because it's so large. I'll look into this and report back.

Can you explain what you mean by "compile time difference could be due to caching?"

nilsvu commented:
@macedo22 re caching: on CI we use ccache to speed up the builds (I think ccache stores compiled object files, and if the source file didn't change it reuses the cached object files instead of compiling). The cache effectiveness varies quite a lot though (e.g. when caches get deleted due to limited storage space on GitHub, or when a low-level header changed so everything down the line gets recompiled), so differences in build time are not necessarily due to changes in the PR.

macedo22 commented:
@nilsvu From what I'm seeing, it seems like things may be crashing because I put inlining back in everywhere in `TensorContract`. When I remove it from a few of the recursive functions, compile time of `Ccz4/Ricci.cpp` goes back to the same amount of compile time and RAM that is used before this PR

I'm going to remove some number of those inlines and benchmark again to see if the removal significantly slows down runtime

macedo22 commented:
@nilsvu Because https://github.com/sxs-collaboration/spectre/pull/3759 needs to be able to update the LHS tensor, I've addressed this in the new fixup that adds an `update` function alongside `evaluate`.

nilsvu commented on tests/Unit/DataStructures/Tensor/Expressions/Test_MixedOperations.cpp:
```diff
@@ -293,12 +428,428 @@ void test_case3(const DataType& used_for_size,
+      for (size_t c = 0; c < Dim + 1; c++) {
+        pi_2_up.get(a, b) += inverse_spacetime_metric.get(c, b) * pi.get(a, c);
+      }
+    }
```
why not use TE here?

macedo22 replied:
In unit tests that test TEs, I avoid using them in computing inputs and expected results since TEs are the thing being tested

nilsvu commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -329,4 +368,58 @@ auto evaluate(const RhsTE& rhs_tensorexpression) {
+ *
+ * One important difference to note with `update` is that it cannot split up the
+ * RHS expression and evaluate subtrees, while `evaluate` can (see
+ * `TensorExpression`). From benchmarking, it was found that the runtime of
```
- Could you add a few comments what the issue with subtrees here is?
- Also, I didn't realize that `evaluate` allows the result tensor as one of the operands in the TE. Is that always safe? E.g., what happens if you compute:

    ```cpp
    evaluate<ti_i, ti_j>(make_not_null(&A), A(ti_j, ti_i));
    ```

    To me it seems nontrivial to make sure permutations like this work correctly with only the result buffer available. Am I wrong, or is workspace memory allocated somewhere?
- Lastly, if update operations like the one above are indeed safe, I think it would be nice to just use `evaluate` for this instead of the extra `update` function (which only calls back into `evaluate` anyway). Is it possible to figure out the `EvaluateSubtrees` flag without the user's help?

If the answer is that these update operations are _not_ safe (or only a subset of them are safe), then I think it's no problem to just have the user compute temporaries themselves in separate variables (workspace memory). It would be good to assert which operations are safe and which are not though.

macedo22 replied:
- Yes, I'll add some comments! The short answer is that if the LHS tensor appears in the RHS expression, we can't use the splitting subtree stuff because that traversal updates the LHS tensor with each subtree that's evaluated (the "accumulation" that's mentioned in the `TensorExpression` class documentation), so a term in RHS would change during the traversal, which is yikes ❌ because the LHS won't be computed correctly.
- You make a good point with the example, which is a case I forgot to consider. If the indices used for the LHS tensor on both the LHS and the RHS are the same, it should be safe. I think I should add a `static_assert` that checks that the generic indices are the same.
- If the generic indices of the LHS tensor used on both sides are the same, then it's safe, but otherwise it's not.

It's possible to eliminate `update`, have `evaluate` just look and see if the LHS tensor is in the RHS expression and, if so, then don't do the split subtree traversal, but the issue is that I believe both tree traversals (the split traversal with `evaluate_primary_subtree`/`get_primary` and the one-liner traversal with `get`) will be compiled because whether or not the LHS tensor is in the RHS expression can only be known at runtime, as opposed to compile time. Compiling both versions isn't ideal because it adds bloat to compile time and in cases where we have a large TE that doesn't have the LHS tensor in the RHS, we definitely wouldn't want the one-liner traversal compiled because `DataVector` expression compile time scales poorly with the number of operations. By having `EvaluateSubtrees`, I think only one of the traversals will be compiled. This was the reasoning for defining a separate function, but let me know if you have follow up thoughts!

macedo22 replied:
- Yes, I'll add some comments! The short answer is that if the LHS tensor appears in the RHS expression, we can't use the splitting subtree stuff because that traversal updates the LHS tensor with each subtree that's evaluated (the "accumulation" that's mentioned in the `TensorExpression` class documentation), so a term in the RHS would change during the traversal, which is yikes ❌ because the LHS won't be computed correctly.
- You make a good point with the example, which is a case I forgot to consider. If the indices used for the LHS tensor on both the LHS and the RHS are the same, it should be safe. I think I should add a `static_assert` that checks that the generic indices are the same.
- If the generic indices of the LHS tensor used on both sides are the same, then it's safe to evaluate it as a one-liner (no splitting), but otherwise it's not safe, like with the example you gave.

It's possible to eliminate `update`, and instead have `evaluate` just look and see if the LHS tensor is in the RHS expression and, if so, then do the one-liner traversal instead of the split subtree traversal. But by having `update`/`EvaluateSubtrees` (something known at compile time), we can ensure at compile time that if people want to use the LHS tensor in the RHS expression, it will be handled safely, as opposed to getting a slap on the wrist at runtime. Moreover, to your point about the example you gave where the indices of the LHS tensor were different on the LHS and RHS, this is similarly something we can `static_assert` not to do at compile time with `update/EvaluateSubtrees`, but only at runtime with `evaluate` if we check for the LHS in the RHS. This was the reasoning for defining a separate function, but let me know if you have follow up thoughts!

macedo22 replied:
Addressed by new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/7adc3cd3837d1eac47e914b3a3fd6f967a53ae21 , which asserts the same index order for LHS used in RHS and documents why it's safe

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -20,6 +20,7 @@
 #include "DataStructures/Tensor/Symmetry.hpp"
```
It might be useful to add a few sentences in the commit message to motivate this rather large set of changes. Just so if someone looks back they'll have an idea of why it's being done :)

macedo22 replied:
(Will do this after squashing all fixups, same for the other commit messages)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -295,34 +295,64 @@ struct AddSub<T1, T2, ArgsList1<Args1...>, ArgsList2<Args2...>, Sign>
                 "Invalid Sign provided for addition or subtraction of Tensor "
```
I think here too it could be useful to expand the commit message with some motivation :)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -265,6 +265,21 @@ struct AddSubType {
 };
```
Again, maybe expand the commit message to add some motivation :)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -618,8 +909,14 @@ SPECTRE_ALWAYS_INLINE auto operator-(
-      "error occurs from expressions like R(ti_a, ti_b) - S(ti_c, ti_a)");
-  return TensorExpressions::AddSub<T1, T2, Args1, Args2, -1>(~t1, ~t2);
+      "error "
+      "occurs from expressions like R(ti_a, ti_b) - S(ti_c, ti_a)");
```
format so `error ` isn't alone on the line? Maybe put the `R - S` expression on its own line instead?

macedo22 replied:
Addressed by new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/AddSubtract.hpp:
```diff
@@ -618,8 +909,14 @@ SPECTRE_ALWAYS_INLINE auto operator-(
+                TensorExpressions::detail::max_num_ops_in_sub_expression<X2>) {
+    return TensorExpressions::AddSub<T1, T2, Args1, Args2, -1>(~t1, ~t2);
+  } else {
+    return -t2 + t1;
```
Sorry, why negate and then add? Doesn't this turn 1 FLOP into 2?

macedo22 replied:
The idea with this is that if `t2` is a TE subtree that contains more than the max # of ops, then move t2 to be the left operand so we can split it up since splitting would be advantageous given that it has "enough" operations to make it worth it. If it's the right operand, it can't be split, as the implementation in this PR only splits along the "leftmost" path from the root to the leftmost leaf. Let me know if you think this is a reasonable thing to do, or not!

macedo22 replied:
Note, however, that while this is beneficial if this swap ends up putting `t2` on this leftmost path, if this swap occurs somewhere off of the leftmost path (i.e. somewhere that won't be split up), then this swap will simply be adding a negation, with no benefit. So, perhaps, it is best to not do any switching if and until I implement splitting elsewhere in the tree, because I don't think it's possible for `operator-()` to know where exactly its operands exist in the tree (i.e. where in the arithmetic expression they exist and if they will or will not be part of the leftmost path)

nilsdeppe replied:
I'm not sure what advice to give you in terms of performance other than measure :) Feel free to leave the PR as-is and measure later! But I think documenting why it's being swapped, or pointing to where it's discussed in other documentation would be helpful here :)

macedo22 replied:
I decided to just remove the swap for now in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Divide.hpp:
```diff
@@ -95,10 +176,114 @@ struct Divide : public TensorExpression<
+  }
+
  private:
-  /// Left operand (numerator)
```
Just double-checking that removing the `(numerator)` and `(denom)` parts was intentional :)

macedo22 replied:
Nope, thanks!

macedo22 replied:
Restored by new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -22,12 +24,231 @@ struct Expression {};
+/// scales poorly as we increase the number of operations. For example, for an
+/// inner product with 256 sums of products, instead of adding 256 `DataVector`
+/// products in one line (e.g. `result = A*B + C*D + E*F + ...;`), it's much
+/// faster to, say, set the result to be the sum of the first 8 products, then
```
maybe add `first 8 (though the exact optimal number will depend on the hardware) products` or a sentence at the end that explains that the optimal size of the subexpressions is hardware-dependent, but probably not something we need to really worry about tuning

macedo22 replied:
Documented in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -70,3 +296,52 @@ template <typename Derived, typename DataType, typename Symm,
+/// `max_num_ops_in_sub_expression_helper`.
+///
+/// Before defining a max operations cap for some data type, the change should
+/// be first justified by benchmarking before and after introducing the new cap.
```
Could you also add here that the optimal cap will likely be hardware-dependent and so if we really wanted to tune the code we would need to benchmark a bunch of different expressions on each hardware architecture and then control the value based on the hardware? I don't think that's something we should spend time on right now, but I think it's a useful thing to note :)

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -22,12 +24,231 @@ struct Expression {};
+/// `Tensor` we're computing does not also appear in the RHS `TensorExpression`.
+/// We do this because if a tree is split up, then the LHS `Tensor` will
+/// generally not be computed correctly due to the fact that the LHS components
+/// will updated as we traverse the split tree.
```
`updated` -> `be updated`?

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -22,12 +24,231 @@ struct Expression {};
+/// const gsl::not_null<LhsTensor*> lhs_tensor) const`: Asserts that the LHS
+/// `Tensor` we're computing does not also appear in the RHS `TensorExpression`.
+/// We do this because if a tree is split up, then the LHS `Tensor` will
+/// generally not be computed correctly due to the fact that the LHS components
```
[nitpick, feel free to ignore] I generally prefer `because` instead of `due to the fact that` :shrug: I think it's more concise and makes the sentence easier to read :)

macedo22 replied:
I enthusiastically accept all pointers on brevity

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -22,12 +24,231 @@ struct Expression {};
+/// - Derived `TensorExpression` classes (or the overloads that produce them)
+/// should include `static_assert`s for ensuring mathematical correctness
+/// wherever reasonable
+/// - Minimize breadth in the tree where possible, as benchmarking inner
```
`as` -> `since`(or `because`) is clearer I think

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -70,3 +296,52 @@ template <typename Derived, typename DataType, typename Symm,
+  // for other data types, define a new static variable like
+  // `max_num_ops_in_datavector_sub_expression` for the data type of interest,
+  // then update the control flow below
+  static constexpr size_t value = std::is_same_v<DataType, DataVector>
```
I think we have an `is_vector_impl` type alias somewhere (I can try to find it if you want). You could use that to generalize the splitting logic to anything that is a Blaze vector we've wrapped :)

macedo22 replied:
Because TEs only work for `double` and `DataVector` right now and because I haven't benchmarked to see if, say, `ComplexDataVector` requires a different number, I think maybe I should keep it specific to `DataVector`, for now. What do you think?

nilsdeppe replied:
Yea, that sounds good to me :)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorIndexTransformation.hpp:
```diff
@@ -151,11 +151,12 @@ transform_multi_index(
   for (size_t i = 0; i < NumIndicesOut; i++) {
     gsl::at(output_multi_index, i) =
-        (gsl::at(tensorindex_transformation, i) ==
+        // Check that the index is not a time index vs that it is, because it is
```
`vs` -> `instead of` and then `because it is` -> `because we expect` (some more fixes to the sentence phrasing to make it real English after my suggestions)

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/e7d1372c7a18d0a4ca4d791e9507327c0621ef87

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -260,6 +237,68 @@ void evaluate(
+
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Evaluate a RHS tensor expression to a tensor with the LHS index order
```
Sorry, I'm having a hard time understanding this brief. Maybe replace `Evaluate` with `Assign` would clarify it. I think the part I don't understand is what it means to "evaluate X to Y" but I do understand what it means to "assign X to Y"

macedo22 replied:
Addressed by new fixups https://github.com/sxs-collaboration/spectre/pull/3842/commits/f4a8355967b0d42817858a11d8927ac71025017c and https://github.com/sxs-collaboration/spectre/pull/3842/commits/05e5f1095b93e4452002d87502e5435d5504caf2

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -260,6 +237,68 @@ void evaluate(
+ * together and fill the provided resultant LHS Tensor `L` with index order
+ * (b, a):
+ * \code{.cpp}
+ * TensorExpressions::evaluate<ti_b, ti_a>(
```
could you point to a snippet in a test? That way at least if any part of the code implementation changes, so does the example :)

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/f4a8355967b0d42817858a11d8927ac71025017c

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -329,4 +368,58 @@ auto evaluate(const RhsTE& rhs_tensorexpression) {
+
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Update a LHS tensor to the result of a RHS tensor expression that
```
Similar comment, the `update X to Y` phrasing doesn't really make it clear to me when I should use this. In a perfect world people will read the entire documentation, but I don't do that and so can't realistically expect others to either :) The brief needs to highlight _when_ this should be used. E.g. `If the LHS tensor is used in the RHS expression, then use this function to compute the result.` or something like that.

macedo22 replied:
Very good point

macedo22 replied:
Addressed by new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/f4a8355967b0d42817858a11d8927ac71025017c

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -329,4 +368,58 @@ auto evaluate(const RhsTE& rhs_tensorexpression) {
+ * \f$R^{i}{}_{jl} S^{l}{}_{i}\f$, followed by a call to `update` to compute the
+ * rest of the expression:
+ * \code{.cpp}
+ * auto L = TensorExpressions::evaluate<ti_j>(
```
could snippet from a test?

macedo22 replied:
Addressed in new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/f4a8355967b0d42817858a11d8927ac71025017c

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -260,6 +237,68 @@ void evaluate(
+ * `TensorExpression` class).
+ *
+ *
+ * Note: `LhsTensorIndices` must be passed by reference because non-type
```
Doxygen has a `\note` functionality.

It would also be good to say why this is safe (assuming it is) because dangling references can be a really painful thing to debug :)

macedo22 replied:
`\note` addressed in new fixup https://github.com/sxs-collaboration/spectre/commit/f4a8355967b0d42817858a11d8927ac71025017c

I'm not sure what you're referring to being safe. If you meant that it should be explained when and why you can use the LHS tensor on the RHS, this is addressed by new fixup https://github.com/sxs-collaboration/spectre/pull/3842/commits/7adc3cd3837d1eac47e914b3a3fd6f967a53ae21

Let me know if you're looking for something else

nilsdeppe replied:
I'm wondering about the `auto&` as the template parameter. I assume it's only used for type info? This comes up because I don't know what the reference points to (it is really a pointer) and so it could be good to clarify that :)

macedo22 replied:
I see. Yes, it's only used for type info.

Can you explain what about it being a reference makes you feel it's worth commenting on why it's safe? For example, is there a common dangling pointer pitfall with reference tparams that you're thinking of and concern I should assuage? I only know using references as tparams is something you can do, but I don't know things like this about them.

I'm asking these things just so I know what dangling pointer danger I should be addressing in the comment and explaining why it's not encountered.

macedo22 replied:
Wait, a guess: are you referring to the possibility of the tparams pointing to a TensorIndex object whose lifetime has ended? If so, then maybe the answer is "we don't have to worry about dangling references because the TensorIndexs you can use are all global static objects" and therefore will never be dangling pointers, unless someone decides to be chaotic and pass their own custom TensorIndex temporaries as tparams. If this unlikely but problematic case of someone using their own custom temporary TensorIndexs is a concern, let me know and I can think about how to prevent people from doing it!

macedo22 replied:
Result frrom Google Meet discussion: no need to worry about dangling pointer because the tparams are global static

nilsdeppe reviewed (changes requested):
Overall looks great, mostly some wording suggestions in documentation :)

macedo22 commented:
Fixups are in! There are quite a few, and I tried to logically compartmentalize them. There are 7 new fixups in total. https://github.com/sxs-collaboration/spectre/pull/3842/commits/2e1af3f1e03bab13d034e5bf3139e5dba03570a6 just fixes some documentation rendering bugs. Other than that, here are the other fixups relevant to your requests, which should also be linked as replies to your comments above:

@nilsvu :
1. https://github.com/sxs-collaboration/spectre/pull/3842/commits/7adc3cd3837d1eac47e914b3a3fd6f967a53ae21
2. https://github.com/sxs-collaboration/spectre/pull/3842/commits/b1566dabe4805a70c10e5a17c698a07774d3fbdb

@nilsdeppe :
1. https://github.com/sxs-collaboration/spectre/pull/3842/commits/904587939fe39bf688e0670b7b21474eada6b315
2. https://github.com/sxs-collaboration/spectre/pull/3842/commits/f4a8355967b0d42817858a11d8927ac71025017c
3. https://github.com/sxs-collaboration/spectre/pull/3842/commits/05e5f1095b93e4452002d87502e5435d5504caf2
4. https://github.com/sxs-collaboration/spectre/pull/3842/commits/e7d1372c7a18d0a4ca4d791e9507327c0621ef87

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -312,7 +318,15 @@ namespace detail {
+///
+/// The current value set for when the data type is `DataVector` was benchmarked
+/// by compiling with clang-10 Release and running on a CentOS Linux 7 system
+/// with a 64-bit x86 architecture.
```
That still leaves around 15 years of different CPUs :) Could you specify the actual CPU? `cat /proc/cpuinfo` should work. You want the `model name`, eg `AMD Ryzen Threadripper 3970X 32-Core Processor` or `Intel(R) Core(TM) i5-4210U CPU @ 1.70GHz`

macedo22 replied:
I got the model name from that command, thanks! Would this change be fine?:
```
... by compiling with clang-10 Release and running on <model name>.
```

Or do you want me to keep the `CentOS Linux 7` and `64-bit x86 architecture` parts, too?

nilsdeppe replied:
I don't think the OS matters so much as the CPU. So if you want to omit the OS, go ahead :)

macedo22 replied:
Squashed in this change

nilsdeppe reviewed (commented):
LGTM, please squash :)

nilsdeppe commented:
Sorry, squash once @nilsvu has looked and said it's okay :P

nilsvu reviewed (dismissed):
Yes please 👍

nilsvu commented:
The gcc-10 Debug build failed and GitHub can't load the log, so I'm not sure what happened. @macedo22 @nilsdeppe how can we make sure this PR is not a regression in compile time and/or memory usage?

macedo22 commented:
@nilsvu I haven't looked at any compile time differences for gcc, so maybe I can try building some things that currently use TEs with gcc-10 and compare

macedo22 commented:
@nilsdeppe I added the CPU, squashed it all, and rebased. Just in case my comment was missed, I'm not 100% clear yet on the dangling pointer request (see [this thread](https://github.com/sxs-collaboration/spectre/pull/3842#discussion_r839992331)). Think this is the last outstanding change request, and then I will take a closer look at compile times

macedo22 commented:
@nilsvu Just wanted to update you that it seems that while gcc-7 and clang-10 Release compile time and RAM hasn't increased by much, the Debug versions' compile time and RAM _has_ by about double for `src/Evolution/Systems/Ccz4/Ricci.cpp`, so I'm currently looking into how to reduce it and will ping you both when I have a fix or more to say

macedo22 commented:
https://github.com/sxs-collaboration/spectre/pull/4023 addressed the compile time issue above, where the compile times/RAM of large TEs in debug mode are now smaller. Since this was merged, I just rebased and added details to the commit messages.

The changes done in the rebase mostly involve updating `TensorExpressions::` -> `tenex::` and `ti_j` -> `ti::j`. I duplicated and saved the branch as it was before rebasing, so let me know if you would like to see the diff with these interface changes.
